### ISSUE TITLE
VFIO VM issues with Zen 5.14.5 kernel on Arch Linux

### ISSUE THREAD

**Issue by JJwA7VR8** [zen-kernel/zen-kernel#245]
I have been using Zen kernel for arch linux host (CPU i7 4790K) for several months. VM is Windows 10 with a GTX 1080 passed through. CPU governor is set to performance in host.

Before 5.14.3, with all CPU cores pinned in VM, I used to get generally smooth performance with intermittent 10-20 second long periods of micro-stutter every few minutes in games like Destiny 2 and Overwatch. Without pinning CPU cores, the FPS would dip noticeably periodically, and overall pinning cores was the better experience.

5.14.3 has been the best zen kernel version for VM performance in my experience, with smooth performance and micro-stutter issues almost completely eliminated in Destiny 2 in the couple hours I tested.

However, 5.14.5 has been the worst zen kernel revision for VM performance. Firstly, I wasn't even able to boot into the VM with CPU cores pinned. After removing all CPU pinning in virt-manager, I was able to boot into the VM, but there were big stutters frequently in Destiny 2 (only game tested as of this version).

I downgraded to zen 5.14.3 using pacman, and the aforementioned good gaming performance with pinned CPU cores is back. So the changes in 5.14.5 from 5.14.3 are problematic as far as VFIO VM gaming are concerned.

**Comment by heftig** [comment]
5.14.5 switched from tweaked CFS to the CacULE+RDB scheduler.

I've also seen Firefox lag terribly while a compile is going on. It's unfortunate it performs so badly, so I'm considering switching back.

**Comment by spica234** [comment]
Well, I tried all schedulers, with my laptop , cpu Ryzen 1st gen 2500U.

I am basically a front end developer, and in my case latency isn't everything that i need but better throughput.

In my testing, I found that Muqss with threadrq on and with rq smt. Provides me the best throughput and latency without compromising each of this. I' m still using latest  5.13 from liquorix source. Tried all other schedulers, I wont say they are bad but for my specific purpose i really miss muqss in 5.14

**Comment by damentz** [comment]
@spica234 have you tried PDS with 2ms scheduling time slices?  This is most likely what Liquorix will switch to for 5.14 unless we get a good port of MuQSS.

**Comment by heftig** [comment]
@JJwA7VR8 Could you test the kernel from https://pkgbuild.com/~heftig/linux-zen-test/ to see if there's any improvement? If this does not help, I will remove CacULE again.

**Comment by spica234** [comment]
> @spica234 have you tried PDS with 2ms scheduling time slices? This is most likely what Liquorix will switch to for 5.14 unless we get a good port of MuQSS.
> 
> If PDS performs noticeably worse than MuQSS, we should bring it to Alfred Chen's attention since he's still developing it.

I will try and let you know for sure.

This is my daily usage where mqss wins:
- I've one docker container live running on my pc
-  other side Visual Studio Code editor for live development, with coding going on side by side along side npm serve going on
- Chrome for constant live debugging with multiple tabs running
- and sometimes on terminal liquorix is compiling with gcc-11
 
With these heavy cpu load, what i found unique with muqss is that my CPU usage is always near 55-60 to sometimes 70-80,  the way these all development stuff is running, that is expected and needed. Muqss is not compromising cpu usage in favor of latency, it's giving sufficient cpu resources, and even if i open any new application it opens smoothly.

There is always  a unique UI smoothness i notice with only muqss.

About other schedulers:
Cacule: It favors latency too much, with all these processes i am daily running it's allocating only 20-30% max cpu, in favor of latency so all my running programs  are taking more time. In early 2010s i was having same better performance with BFS as well. and during my xda development BFS was performing better as well on android in early days of android.

Anyways, i will update you with PDS by setting slice period suggested by you.

**Comment by heftig** [comment]
> @JJwA7VR8 Could you test the kernel from https://pkgbuild.com/~heftig/linux-zen-test/ to see if there's any improvement? If this does not help, I will remove CacULE again.

The -zen2 kernel in this test folder has the "basic" CacULE without RDB, but the -zen2 kernel I just released to [extra] has CacULE completely removed. This issue should now be fixed, although feedback for the test kernel would still be welcome.

**Comment by JJwA7VR8** [comment]
> > @JJwA7VR8 Could you test the kernel from https://pkgbuild.com/~heftig/linux-zen-test/ to see if there's any improvement? If this does not help, I will remove CacULE again.
> 
> The -zen2 kernel in this test folder has the "basic" CacULE without RDB, but the -zen2 kernel I just released to [extra] has CacULE completely removed. This issue should now be fixed, although feedback for the test kernel would still be welcome.

I did the following set of quick testing from the perspective of running a GPU passthrough VM, I didn't test responsiveness, performance for the host itself...

First to make sure that the issue in 5.14.5 was due to linux-zen changes and not due to anything from the vanilla kernel, I booted up the vanilla arch kernel 5.14.5 and ran the VM.
- VM booted with pinned cores just fine, performance was smooth in Destiny 2 with one occurrence of slight micro-stutter, but overall fine. Average FPS seemed to be perhaps a couple FPS lower than usual zen kernel FPS (on a base of ~ average 80+ fps) but due to absence of canned benchmark I cannot say with certainty.

Secondly, I manually downloaded and installed the 5.14.5 zen2-1 test kernel via `pacman -U`
- Like the original 5.14.5 zen1-1 kernel from repo, I could not boot into the VM with pinned cores using the test kernel.
- Removed pinning in virt-manager, and the VM booted just fine. However unlike the original 5.14.5 zen1-1 kernel from repo, there were no massive or frequent stutters. The frame rate was generally smooth, but there were noticeable dips and overall performance seemed lower than usual zen kernel performance by 5 fps (again on a base of ~ 80+ fps)
- I also ran Cinebench R20 and scored a lower than expected avg score of 1912 over two runs.

Finally, I installed the updated 5.14.5 zen2-1 kernel using pacman upgrade.
- VM booted up just fine with pinned cores like with 5.14.3
- Destiny 2 performance  seems back up to 5.14.3 levels
- Cinebench scored ~1940 with pinned cores which is higher than the unpinned core performance of test kernel. Usually, pinning cores actually lowers the cinebench score as compared to unpinned cores (while reducing latency)
